### PR TITLE
Add LDAP authentication

### PR DESCRIPTION
**Pull request by DoubleMalt**
This pul request enables LDAP authentication. If a users authenticate successfully via the configured LDAP server and are not yet in the local database, they are created.

Missing parts:
- get email from LDAP server
- get full name from LDAP server

### CODE REVIEW

**Comment by erikjohnston** [comment]
@matrixbot ok to test

(Sorry about the matrixbot spam)

**Comment by erikjohnston** [comment]
Oh, I'll need to add the new dependency manually to get the tests to run.

**Comment by erikjohnston** [comment]
Although see: http://matrix.org/jenkins/job/SynapseFlake8Packaging/176/violations/file/synapse/handlers/auth.py/ for some code style violations :)

File: synapse/handlers/auth.py [hunk 1]
```diff
@@ -407,11 +423,51 @@ def _find_user_id_and_pwd_hash(self, user_id):
...
             defer.returnValue(user_infos.popitem())
 
-    def _check_password(self, user_id, password, stored_hash):
-        """Checks that user_id has passed password, raises LoginError if not."""
-        if not self.validate_hash(password, stored_hash):
-            logger.warn("Failed password login for user %s", user_id)
-            raise LoginError(403, "", errcode=Codes.FORBIDDEN)
+    @defer.inlineCallbacks
+    def _check_password(self, user_id, password):
+        defer.returnValue(not ((yield self._check_ldap_password(user_id, password)) or (yield self._check_local_password(user_id, password))))
```
Comment (erikjohnston) [inline synapse/handlers/auth.py]: This is a bit hard to follow. Maybe something like the following is clearer?

```python
if self.ldap_enabled:
    valid_ldap = yield self._check_ldap_password(user_id, password))
    if valid_ldap:
       defer.returnValue(True)

valid_password = yield self._check_local_password(user_id, password)
defer.returnValue(valid_password)
```

This way we don't attempt (and log) if we've tried to do LDAP each time someone logs in to a HS with LDAP disabled.

File: synapse/handlers/auth.py [hunk 2]
```diff
@@ -407,11 +423,51 @@ def _find_user_id_and_pwd_hash(self, user_id):
...
+
+
+    @defer.inlineCallbacks
+    def _check_local_password(self, user_id, password):
+        try:
+            user_id, password_hash = yield self._find_user_id_and_pwd_hash(user_id)
+            defer.returnValue(not self.validate_hash(password, password_hash))
+        except:
```
Comment (erikjohnston) [inline synapse/handlers/auth.py]: I'd prefer if this only caught login failures, as otherwise it masks programmer/logic errors.

File: synapse/handlers/auth.py [hunk 3]
```diff
@@ -407,11 +423,51 @@ def _find_user_id_and_pwd_hash(self, user_id):
...
+            l.simple_bind_s(dn.encode('utf-8'), password.encode('utf-8'))
+
+            if not (yield self.does_user_exist(user_id)):
+                user_id, access_token = (
+                    yield self.hs.get_handlers().registration_handler.register(localpart=local_name)
+                )
+
+            defer.returnValue(True)
+        except ldap.LDAPError, e:
+            logger.info("LDAP error: %s" % e)
```
Comment (erikjohnston) [inline synapse/handlers/auth.py]: Maybe make this a warning? Also, instead of string formatting you can do: `logger.info("LDAP error: %s", e)`, which saves us doing unnecessary string formatting.

File: synapse/handlers/auth.py (hunk 1 above)
Comment (DoubleMalt) [inline synapse/handlers/auth.py]: the first line of _check_ldap_password bails immediately if ldap_enabled is false. I can put the log level for the message on debug.

I honestly have a harder time parsing your code to see that it returns true if any of the two methods return true, but that's probably just my personal preference. I don't have a problem to change the code to you suggestion if you feel strongly about it :)

**Comment by erikjohnston** [comment]
Does `python-ldap` have a system dependency? If it does then I'd probably want to make the `python-ldap` dependency optional somehow, otherwise its going to break everyone's synapse on upgrade.

File: synapse/handlers/auth.py (hunk 2 above)
Comment (DoubleMalt) [inline synapse/handlers/auth.py]: Good point. Will fix.

**Comment by DoubleMalt** [comment]
Unfortunately it depends on openldap-dev. How could I make it optional in the python-dependencies file?

**Comment by erikjohnston** [comment]
> Unfortunately it depends on openldap-dev. How could I make it optional in the python-dependencies file?

I think the easiest is just to omit it from the dependency list entirely, and only try and import it if ldap is enabled. Since importing is cheap after its done initially, I'd probably import it in `__init__` (to catch early if ldap is enabled, but they didn't add the dependency) and then just reimport it in `_check_ldap_password`

**Comment by erikjohnston** [comment]
The unit tests are failing due to the fact that Mocks evaluate to True. I propose being slightly evil and changing the ldap enabled tests to `if self.ldap_enabled is True` and `if self.ldap_enabled is not True`

File: synapse/handlers/auth.py [hunk 4]
```diff
@@ -407,11 +425,60 @@ def _find_user_id_and_pwd_hash(self, user_id):
...
-            logger.warn("Failed password login for user %s", user_id)
-            raise LoginError(403, "", errcode=Codes.FORBIDDEN)
+    @defer.inlineCallbacks
+    def _check_password(self, user_id, password):
+        defer.returnValue(
+            not (
+                (yield self._check_ldap_password(user_id, password))
+                or
+                (yield self._check_local_password(user_id, password))
+            ))
```
Comment (erikjohnston) [inline synapse/handlers/auth.py]: Line 234 needs to be updated now that this no longer raises but instead returns a boolean.

**Comment by DoubleMalt** [comment]
All checks pass :)

**Comment by erikjohnston** [comment]
Excellent! If you could just sign off on this contribution we can land it: https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off

Simply replying here with `Signed-off-by: Your Name <your@email.example.org>`, is sufficient.

Also, feel free to add yourself to the `AUTHORS.rst`!

**Comment by erikjohnston** [comment]
Woo! Thanks :)

(The failure seems to be due to jenkins failing to set the commit status on github, intriguingly. All the tests pass though)

**Comment by aperezdc** [comment]
JFTR if you want to avoid the dependency of OpenLDAP installed in the system, a pure-Python option would be the [ldap3](https://pypi.python.org/pypi/ldap3/) package.
